### ISSUE TITLE
Question about functional.reset_net

### ISSUE THREAD

**Issue by GorkaAbad** [fangwei123456/spikingjelly#318]
Hi,
I wanted to ask when is desirable to use ```functional.reset_net```. 
More precisely, I want to train a model twice in the same training loop, so I was wondering if functional.reset_net(model) should be called after every usage of ```out = model(x)``` or after each step of the optimizer.

```python
for x,y in trainset:
    out = model(x)
    # (1) Use functional.reset_net(model) here
    loss = loss_fn(out, y)
    loss.backward()
    optimizer.step()
    # (2) Use functional.reset_net(model) here
```

Which of the two approaches is the desirable one?

Thanks in advance,
Gorka

**Comment by fangwei123456** [comment]
Hi, it is a good question.
We recommend to call `reset_net` after `optimizer.step()`.

For example, if your loss depends on `.v` of the neuron, and you call `loss = loss_fn(model[-1].v, target)` , then you will get a wrong loss because `model[-1].v` is resetted to `0` (if using hard reset).

**Comment by GorkaAbad** [comment]
Thanks for the fast reply,
Based on your answer, the following code will be wrong, right?
```python
for x,y in trainset:
    out = model(x)
    out_2 = model(x')
    loss = loss_fn(out, y)
    loss_2 = loss_fn(out_2, y')    
    loss.backward()
    loss_2.backward()
    optimizer.step()
    optimizer_2.step()
    # Use functional.reset_net(model) here
```

The ```out_2``` will be wrong because we have not done the reset yet (I assume). So, maybe in this case is preferable to call ```reset_net``` after ```out = model(x)```.
Am I right?

Thanks in advance,
Gorka

**Comment by fangwei123456** [comment]
It depends on your purpose. If `x` and `x'` are independ samples, then you should not call `model(x')` before reseting model's hidden states.

> The `out_2` will be wrong because we have not done the reset yet (I assume).

Yes, because your model is using hidden states created from `x`.

**Comment by GorkaAbad** [comment]
It is crystal clear now!
Thanks!
